File: backend-spring-boot/src/main/java/no/vicx/backend/user/validation/RecaptchaThenUniqueUsername.java
package no.vicx.backend.user.validation;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Constraint(validatedBy = RecaptchaThenUniqueUsernameValidator.class)
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
public @interface RecaptchaThenUniqueUsername {
    String message() default "";
    String recaptchaMessage() default "{vicx.constraints.reCAPTCHA.message}";
    int usernameMinLength() default 4;
    String uniqueUsernameMessage() default "{vicx.constraints.username.UniqueUsername.message}";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: backend-spring-boot/src/main/kotlin/no/vicx/backend/user/validation/RecaptchaThenUniqueUsername.kt
package no.vicx.backend.user.validation

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass


@Constraint(validatedBy = [RecaptchaThenUniqueUsernameValidator::class])
@Target(AnnotationTarget.CLASS)
@Retention(
    AnnotationRetention.RUNTIME
)
annotation class RecaptchaThenUniqueUsername(
    val message: String = "",
    val recaptchaMessage: String = "{vicx.constraints.reCAPTCHA.message}",
    val usernameMinLength: Int = 4,
    val uniqueUsernameMessage: String = "{vicx.constraints.username.UniqueUsername.message}",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)
